document/document.go

```diff
@@ -86,16 +86,107 @@ func (fb FieldBuffer) GetByField(field string) (Value, error) {
 	return Value{}, ErrFieldNotFound
 }
 
+// setFieldValue replaces a field if it already exists or creates one if not.
+func (fb *FieldBuffer) setFieldValue(field string, reqValue Value) error {
+	_, err := fb.GetByField(field)
+	switch err {
+	case ErrFieldNotFound:
+		fb.Add(field, reqValue)
+		return nil
+	case nil:
+		_ = fb.Replace(field, reqValue)
+		return nil
+	}
+
+	return err
+}
+
+// setValueAtPath deep replaces or creates a field
+// through the value path to get the value and create or replace it.
+func (fb *FieldBuffer) setValueAtPath(v Value, p ValuePath, newValue Value) (Value, error) {
+
+	switch v.Type {
+	case DocumentValue:
+		var buf FieldBuffer
+		err := buf.ScanDocument(v.V.(Document))
+		if err != nil {
+			return v, err
+		}
+
+		if len(p) == 1 {
+			err = buf.setFieldValue(p[0], newValue)
+			return NewDocumentValue(&buf), err
+		}
+
+		va, err := buf.GetByField(p[0])
+		if err != nil {
+			return v, err
+		}
+
+		va, err = buf.setValueAtPath(va, p[1:], newValue)
+		if err != nil {
+			return v, err
+		}
+
+		err = buf.setFieldValue(p[0], va)
+		return NewDocumentValue(&buf), err
+	case ArrayValue:
+		var vb ValueBuffer
+		err := vb.ScanArray(v.V.(Array))
+		if err != nil {
+			return v, err
+		}
+
+		index, err := strconv.Atoi(p[0])
+		if err != nil {
+			return v, err
+		}
+
+		va, err := vb.GetByIndex(index)
+		if err != nil {
+			return v, err
+		}
+
+		if len(p) == 1 {
+			err = vb.Replace(index, newValue)
+			return NewArrayValue(&vb), err
+		}
+
+		va, err = fb.setValueAtPath(va, p[1:], newValue)
+		err = vb.Replace(index, va)
+		return NewArrayValue(&vb), err
+	}
+
+	return v, nil
+}
+
 // Set replaces a field if it already exists or creates one if not.
-func (fb *FieldBuffer) Set(f string, v Value) {
-	for i := range fb.fields {
-		if fb.fields[i].Field == f {
+func (fb *FieldBuffer) Set(path ValuePath, reqValue Value) error {
+	// check if the ValuePath contains only one field to set or replace
+	if len(path) == 1 {
+		// Set or replace the unique field
+		return fb.setFieldValue(path[0], reqValue)
+	}
+
+	for i, field := range fb.fields {
+		if path[0] == field.Field {
+			var buf FieldBuffer
```

Comment:
Why do we need to deep copy `fb`, since it's already a FieldBuffer? Is there a reason why we don't modify it directly?

Reply:
You 're right